apps/microsoft-teams/app-actions/src/actions/list-channels.ts

```diff
@@ -1,16 +1,54 @@
 import { AppActionCallContext } from '@contentful/node-apps-toolkit';
 import { AppActionCallResponse, Channel } from '../types';
-import { mockChannels } from '../../test/fixtures/mockChannels';
-
-interface AppActionCallParameters {
-  tenantId: string;
-}
+import { fetchTenantId } from '../utils';
+import helpers from '../helpers';
+import { config } from '../config';
 
 export const handler = async (
-  _payload: AppActionCallParameters,
-  _context: AppActionCallContext
+  _payload: {},
+  context: AppActionCallContext
 ): Promise<AppActionCallResponse<Channel[]>> => {
-  const channels = mockChannels;
+  const {
+    cma,
+    appActionCallContext: { appInstallationId },
+  } = context;
+
+  let channels: Channel[];
+
+  try {
+    if (config.botServiceUrl === undefined) {
+      throw new Error('MS Teams Bot Service URL not provided.');
+    }
+
+    if (config.apiKey === undefined) {
+      throw new Error('MS Teams Bot Service API Key not provided.');
+    }
+
+    const tenantId = await fetchTenantId(cma, appInstallationId);
+    channels = await helpers.getChannelsList(config.botServiceUrl, config.apiKey, tenantId);
```

Comment:
I think this is okay for now but to promote reuse I actually recommend we adopt the "client pattern" (not an official term, i just made that up) here.

basically instead of buiding ad hoc fetch requests every time we want to pull data from our MS Teams service we should build a dedicated client class around that API and implement each different kind of request as a method.

this also makes it much easier to test and later to evolve.

Reply:
I created a follow up ticket for us to address this refactoring later!